`Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference` error in the controller pod

## Bug information
I have observed this error message in the functional test runs of some PRs in the controller pod.

`
{"severity":"info","timestamp":"2023-11-22T16:40:35.409Z","name":"controller.controller","caller":"controller/controller.go:114","message":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","serviceName":"controller","version":"edge","hostName":"controller-84f587c5d6-kqhcf","controller":"deployment","controllerGroup":"apps","controllerKind":"Deployment","Deployment":{"name":"dapr-sts-manual-redis","namespace":"default-daprrp-rs-statestore-manual"},"namespace":"default-daprrp-rs-statestore-manual","name":"dapr-sts-manual-redis","reconcileID":"276efd5d-463d-41b1-8b3c-30c5312ed1fd"}
`

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x17b5417]

goroutine 133 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115 +0x1e5
panic({0x1974860?, 0x2cda2d0?})
	/opt/hostedtoolcache/go/1.21.4/x64/src/runtime/panic.go:914 +0x21f
github.com/radius-project/radius/pkg/controller/reconciler.(*DeploymentReconciler).Reconcile(0xc0000cb100, {0x1ed8998, 0xc000908660}, {{{0xc00095dcb0?, 0x0?}, {0xc000691068?, 0x410745?}}})
	/home/runner/work/radius/radius/pkg/controller/reconciler/deployment_reconciler.go:110 +0x3b7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1ed8998?, {0x1ed8998?, 0xc000908660?}, {{{0xc00095dcb0?, 0x18a9e60?}, {0xc000691068?, 0x1ec7010?}}})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001ad040, {0x1ed89d0, 0xc0006ec870}, {0x1a1fb00?, 0xc0005b78c0?})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314 +0x365
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001ad040, {0x1ed89d0, 0xc0006ec870})
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 101
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222 +0x565

### Steps to reproduce (required)
1. Create a PR or run functional tests locally (the latter needs confirmation)
2. Check controller pod logs
3. See the panic

### Observed behavior (required)
Error in the controller pod.

### Desired behavior (required)
No panic

### Workaround (optional)

## System information

### rad Version (required)
PR

### Operating system (required)
PR

## Additional context
[msgrp_container_logs.zip](https://github.com/radius-project/radius/files/13443632/msgrp_container_logs.zip)
[daprrp_container_logs.zip](https://github.com/radius-project/radius/files/13443633/daprrp_container_logs.zip)



[AB#10470](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/10470)

--- a/pkg/controller/reconciler/annotations.go
+++ b/pkg/controller/reconciler/annotations.go
@@ -82,9 +82,9 @@ type deploymentStatus struct {
 // readAnnotations reads the annotations from a Deployment.
 //
 // This includes the configuration specified by the user, the hash of the configuration, and the status.
-func readAnnotations(deployment *appsv1.Deployment) (*deploymentAnnotations, error) {
+func readAnnotations(deployment *appsv1.Deployment) (deploymentAnnotations, error) {
 	if deployment.Annotations == nil {
-		return nil, nil
+		return deploymentAnnotations{}, nil
 	}
 
 	result := deploymentAnnotations{
@@ -96,7 +96,7 @@ func readAnnotations(deployment *appsv1.Deployment) (*deploymentAnnotations, err
 	if status != "" {
 		err := json.Unmarshal([]byte(status), &s)
 		if err != nil {
-			return &result, fmt.Errorf("failed to unmarshal status annotation: %w", err)
+			return result, fmt.Errorf("failed to unmarshal status annotation: %w", err)
 		}
 	}
 
@@ -106,7 +106,7 @@ func readAnnotations(deployment *appsv1.Deployment) (*deploymentAnnotations, err
 	// This is important so that can clean up previously created connections when Radius is disabled.
 	enabled := deployment.Annotations[AnnotationRadiusEnabled]
 	if !strings.EqualFold(enabled, "true") {
-		return &result, nil
+		return result, nil
 	}
 
 	result.Configuration = &deploymentConfiguration{
@@ -121,7 +121,7 @@ func readAnnotations(deployment *appsv1.Deployment) (*deploymentAnnotations, err
 		}
 	}
 
-	return &result, nil
+	return result, nil
 }
 
 // ApplyToDeployment applies the configuration and status to a Deployment.

--- a/pkg/controller/reconciler/deployment_reconciler.go
+++ b/pkg/controller/reconciler/deployment_reconciler.go
@@ -107,7 +107,7 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 		// and repair it on the next reconcile.
 	}
 
-	if annotations.Status.Operation != nil {
+	if annotations.Status != nil && annotations.Status.Operation != nil {
 		// NOTE: if reconcileOperation completes successfully, then it will return a "zero" result,
 		// this means the operation has completed and we should continue processing.
 		result, err := r.reconcileOperation(ctx, &deployment, annotations)
@@ -181,7 +181,6 @@ func (r *DeploymentReconciler) reconcileOperation(ctx context.Context, deploymen
 		annotations.Status.Operation = nil
 		annotations.Status.Container = annotations.Status.Scope + "/providers/Applications.Core/containers/" + deployment.Name
 		return ctrl.Result{}, nil
-
 	} else if annotations.Status.Operation.OperationKind == radappiov1alpha3.OperationKindDelete {
 		poller, err := r.Radius.Containers(annotations.Status.Scope).ContinueDeleteOperation(ctx, annotations.Status.Operation.ResumeToken)
 		if err != nil {

--- a/pkg/controller/reconciler/deployment_reconciler.go
+++ b/pkg/controller/reconciler/deployment_reconciler.go
@@ -110,7 +110,7 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 	if annotations.Status != nil && annotations.Status.Operation != nil {
 		// NOTE: if reconcileOperation completes successfully, then it will return a "zero" result,
 		// this means the operation has completed and we should continue processing.
-		result, err := r.reconcileOperation(ctx, &deployment, annotations)
+		result, err := r.reconcileOperation(ctx, &deployment, &annotations)
 		if err != nil {
 			logger.Error(err, "Unable to reconcile in-progress operation.")
 			return ctrl.Result{}, err
@@ -127,10 +127,10 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 	// If the Deployment is being deleted **or** if Radius is no longer enabled, then we should
 	// clean up any Radius state.
 	if deployment.DeletionTimestamp != nil || (annotations.Configuration == nil && annotations.Status != nil) {
-		return r.reconcileDelete(ctx, &deployment, annotations)
+		return r.reconcileDelete(ctx, &deployment, &annotations)
 	}
 
-	return r.reconcileUpdate(ctx, &deployment, annotations)
+	return r.reconcileUpdate(ctx, &deployment, &annotations)
 }
 
 // reconcileOperation reconciles a Deployment that has an operation in progress.
@@ -249,12 +249,12 @@ func (r *DeploymentReconciler) reconcileUpdate(ctx context.Context, deployment *
 	}
 
 	environmentName := "default"
-	if annotations.Configuration.Environment != "" {
+	if annotations.Configuration != nil && annotations.Configuration.Environment != "" {
 		environmentName = annotations.Configuration.Environment
 	}
 
 	applicationName := deployment.Namespace
-	if annotations.Configuration.Application != "" {
+	if annotations.Configuration != nil && annotations.Configuration.Application != "" {
 		applicationName = annotations.Configuration.Application
 	}
 
@@ -263,6 +263,10 @@ func (r *DeploymentReconciler) reconcileUpdate(ctx context.Context, deployment *
 		r.EventRecorder.Event(deployment, corev1.EventTypeWarning, "DependencyError", err.Error())
 		logger.Error(err, "Unable to resolve dependencies.")
 		return ctrl.Result{}, fmt.Errorf("failed to resolve dependencies: %w", err)
+	}
+
+	if annotations.Status == nil {
+		annotations.Status = &deploymentStatus{}
 	}
 
 	annotations.Status.Scope = resourceGroupID
@@ -654,7 +658,7 @@ func (r *DeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		annotations, err := readAnnotations(deployment)
 		if err != nil {
 			return []string{}
-		} else if annotations == nil || annotations.Configuration == nil {
+		} else if annotations.Configuration == nil {
 			return []string{}
 		}
 

--- a/pkg/controller/reconciler/deployment_reconciler_test.go
+++ b/pkg/controller/reconciler/deployment_reconciler_test.go
@@ -482,7 +482,7 @@ func waitForStateWaiting(t *testing.T, client client.Client, name types.Namespac
 	ctx := testcontext.New(t)
 
 	logger := t
-	var annotations *deploymentAnnotations
+	var annotations deploymentAnnotations
 	require.EventuallyWithTf(t, func(t *assert.CollectT) {
 		logger.Logf("Fetching Deployment: %+v", name)
 		current := &appsv1.Deployment{}
@@ -499,14 +499,14 @@ func waitForStateWaiting(t *testing.T, client client.Client, name types.Namespac
 		}
 	}, deploymentTestWaitDuration, deploymentTestWaitInterval, "waiting for state to be Waiting")
 
-	return annotations
+	return &annotations
 }
 
 func waitForStateUpdating(t *testing.T, client client.Client, name types.NamespacedName) *deploymentAnnotations {
 	ctx := testcontext.New(t)
 
 	logger := t
-	var annotations *deploymentAnnotations
+	var annotations deploymentAnnotations
 	require.EventuallyWithTf(t, func(t *assert.CollectT) {
 		logger.Logf("Fetching Deployment: %+v", name)
 		current := &appsv1.Deployment{}
@@ -523,14 +523,14 @@ func waitForStateUpdating(t *testing.T, client client.Client, name types.Namespa
 		}
 	}, deploymentTestWaitDuration, deploymentTestWaitInterval, "waiting for state to be Updating")
 
-	return annotations
+	return &annotations
 }
 
 func waitForStateReady(t *testing.T, client client.Client, name types.NamespacedName) *deploymentAnnotations {
 	ctx := testcontext.New(t)
 
 	logger := t
-	var annotations *deploymentAnnotations
+	var annotations deploymentAnnotations
 	require.EventuallyWithTf(t, func(t *assert.CollectT) {
 		logger.Logf("Fetching Deployment: %+v", name)
 		current := &appsv1.Deployment{}
@@ -547,14 +547,14 @@ func waitForStateReady(t *testing.T, client client.Client, name types.Namespaced
 		}
 	}, deploymentTestWaitDuration, deploymentTestWaitInterval, "waiting for state to be Ready")
 
-	return annotations
+	return &annotations
 }
 
 func waitForStateDeleting(t *testing.T, client client.Client, name types.NamespacedName) *deploymentAnnotations {
 	ctx := testcontext.New(t)
 
 	logger := t
-	var annotations *deploymentAnnotations
+	var annotations deploymentAnnotations
 	require.EventuallyWithTf(t, func(t *assert.CollectT) {
 		logger.Logf("Fetching Deployment: %+v", name)
 		current := &appsv1.Deployment{}
@@ -571,7 +571,7 @@ func waitForStateDeleting(t *testing.T, client client.Client, name types.Namespa
 		}
 	}, deploymentTestWaitDuration, deploymentTestWaitInterval, "waiting for state to be Deleting")
 
-	return annotations
+	return &annotations
 }
 
 func waitForRadiusContainerDeleted(t *testing.T, client client.Client, name types.NamespacedName) *deploymentAnnotations {

